### PR TITLE
[Android] store borderColor in a non lossy way

### PR DESCRIPTION
**Pull request by manicakes**
Fix for issue #3652 -

Converting from `int` to `float` is lossy for very large numbers, so storing `borderColor` as a single `Spacing` object (which uses `float`) was not working for certain colors.

So, this pull request splits `borderColor` into alpha and RGB components, and stores each of these as their own respective `Spacing` objects.

*Test Plan*

Check out @cosmith sample code here that triggers the bug: https://rnplay.org/apps/l1bw2A

What currently looks like this:

screen shot 2016-08-13 at 6 22 28 pm

Should look like this (with my fix applied):

screen shot 2016-08-13 at 6 20 08 pm

### CODE REVIEW

**Comment by manicakes** [comment]
My fix currently causes rendering issues. Needs more work.

**Comment by facebook-github-bot-3** [comment]
@manicakes updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@manicakes updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@manicakes updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@manicakes updated the pull request.

**Comment by manicakes** [comment]
@andreicoman11 and @foghina looking forward to your feedback - thanks!

**Comment by facebook-github-bot-9** [comment]
@manicakes updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@manicakes updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@manicakes updated the pull request.

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java [hunk 1]
```diff
@@ -46,6 +46,8 @@
 /* package */ class ReactViewBackgroundDrawable extends Drawable {
 
   private static final int DEFAULT_BORDER_COLOR = Color.BLACK;
```
Comment (foghina) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java]: Is this still used anywhere? Can we remove it?

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java [hunk 2]
```diff
@@ -74,6 +76,7 @@
   /* Value at Spacing.ALL index used for rounded borders, whole array used by rectangular borders */
   private @Nullable Spacing mBorderWidth;
   private @Nullable Spacing mBorderColor;
```
Comment (foghina) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java]: Rename to `mBorderRgb`?

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java [hunk 3]
```diff
@@ -164,16 +167,32 @@ public void setBorderWidth(int position, float width) {
     }
   }
 
-  public void setBorderColor(int position, float color) {
+  public void setBorderRGB(int position, float rgb) {
```
Comment (foghina) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java]: We should combine this and `setBorderAlpha` into a single public-facing API, e.g. `setBorderColor(int position, float rgb, float alpha)`, I don't think calling one without the other would make any sense and hints at a programming error. Feel free to keep the methods separate but make them private.

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java [hunk 4]
```diff
@@ -201,8 +201,12 @@ public void setBorderWidth(int position, float width) {
     getOrCreateReactViewBackground().setBorderWidth(position, width);
   }
 
-  public void setBorderColor(int position, float color) {
-    getOrCreateReactViewBackground().setBorderColor(position, color);
+  public void setBorderRGB(int position, float rgb) {
```
Comment (foghina) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: Same as above.

**Comment by foghina** [comment]
Sweet, thanks for picking this up! Can you also provide a test plan for your PR? E.g. a screenshot of the sample app from #3652 running correctly.

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java (hunk 3 above)
Comment (manicakes) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java]: Good idea - will do.

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java (hunk 4 above)
Comment (manicakes) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: 👍

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java (hunk 2 above)
Comment (manicakes) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java]: 👍

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java (hunk 1 above)
Comment (manicakes) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java]: It's used in the subsequent lines, as I split out the rgb and alpha components. Hope that makes sense.

**Comment by manicakes** [comment]
@foghina no prob! Will include test plan and update the PR shortly.

**Comment by facebook-github-bot-5** [comment]
@manicakes updated the pull request.

**Comment by manicakes** [comment]
@foghina updated - thanks for the feedback.

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java (hunk 1 above)
Comment (foghina) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java]: Ah, I must be blind. 👓

**Comment by foghina** [comment]
@facebook-github-bot shipit

awesomesauce

**Comment by facebook-github-bot-9** [comment]
@manicakes updated the pull request.

**Comment by foghina** [comment]
Hmm, something didn't work. Lemme try again.

@facebook-github-bot shipit pls

**Comment by manicakes** [comment]
Hmmm @foghina maybe the shipit command has to be the only content in the comment for it to work? :shipit:

**Comment by foghina** [comment]
It usually works with other stuff in there as well, but yeah lemme try.

**Comment by foghina** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-3** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3716707) to review internal test results.

**Comment by foghina** [comment]
Yay. We should make "shipit pls" a command though.

**Pull request closed without merging**